File: app/src/main/java/org/cerion/stockcharts/positions/PortfolioPieChart.java
package org.cerion.stockcharts.positions;

import android.content.Context;
import android.graphics.Color;
import androidx.annotation.Nullable;
import android.util.AttributeSet;
import android.util.Pair;

import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.data.PieEntry;
import com.github.mikephil.charting.formatter.PercentFormatter;

import java.util.ArrayList;
import java.util.List;

public class PortfolioPieChart extends PieChart {

    //PieChart chart;
    public PortfolioPieChart(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        getLegend().setEnabled(false);
        setDescription(null);
        setEntryLabelColor(Color.BLACK);
    }

    public void setAllocations(List<Pair<String, Float>> allocations) {
        float total = 0;
        for(int i = 0; i < allocations.size(); i++){
            Pair<String,Float> a = allocations.get(i);
            total += a.second;
        }

        ArrayList<PieEntry> entries = new ArrayList<>();
        for(int i = 0; i < allocations.size(); i++){
            Pair<String,Float> a = allocations.get(i);
            entries.add(new PieEntry(100 * a.second / total, a.first));
        }

        // Data set
        PieDataSet dataSet = new PieDataSet(entries, "");
        dataSet.setSliceSpace(2);
        dataSet.setValueTextSize(12);
        dataSet.setValueFormatter(new PercentFormatter());

        //Colors
        ArrayList<Integer> colors = new ArrayList<>();
        colors.add(Color.GRAY);
        colors.add(Color.BLUE);
        colors.add(Color.RED);
        colors.add(Color.GREEN);
        colors.add(Color.CYAN);
        colors.add(Color.YELLOW);
        colors.add(Color.MAGENTA);
        dataSet.setColors(colors);

        PieData pieData = new PieData(dataSet);
        setData(pieData);
        invalidate();
    }
}

File: app/src/main/java/org/cerion/stockcharts/positions/PortfolioPieChart.kt
package org.cerion.stockcharts.positions

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.util.Pair
import com.github.mikephil.charting.charts.PieChart
import com.github.mikephil.charting.data.PieData
import com.github.mikephil.charting.data.PieDataSet
import com.github.mikephil.charting.data.PieEntry
import com.github.mikephil.charting.formatter.PercentFormatter
import org.cerion.stocks.core.model.Position
import java.util.*

class PortfolioPieChart(context: Context?, attrs: AttributeSet?) : PieChart(context, attrs) {

    init {
        legend.isEnabled = false
        description = null
        setEntryLabelColor(Color.BLACK)
    }

    fun setPositions(positions: List<Position>) {
        val total = positions.sumByDouble { it.totalValue }.toFloat() / 100

        val entries = ArrayList<PieEntry>()
        val colors = ArrayList<Int>()

        val cashPosition = positions.firstOrNull { it.cash }
        if (cashPosition != null) {
            // TODO merge or add multiple colors
            colors.add(Color.GREEN)
            entries.add(PieEntry(cashPosition.totalValue.toFloat() / total, "Cash"))
        }

        for (p in positions.filter { !it.cash }) {
            val value = (p.totalValue / total).toFloat()
            val label = if(p.cash) "Cash" else p.symbol
            entries.add(PieEntry(value, label))
        }

        // Data set
        val dataSet = PieDataSet(entries, "")
        dataSet.sliceSpace = 2f
        dataSet.valueTextSize = 12f
        dataSet.valueFormatter = PercentFormatter()

        colors.add(Color.GRAY)
        colors.add(Color.BLUE)
        colors.add(Color.RED)
        colors.add(Color.CYAN)
        colors.add(Color.YELLOW)
        colors.add(Color.MAGENTA)
        dataSet.colors = colors
        val pieData = PieData(dataSet)
        data = pieData
        invalidate()
    }

    @Deprecated(message = "use positions instead")
    fun setAllocations(allocations: List<Pair<String?, Float>>) {
        var total = 0f
        for (i in allocations.indices) {
            val a = allocations[i]
            total += a.second
        }
        val entries = ArrayList<PieEntry>()
        for (i in allocations.indices) {
            val a = allocations[i]
            entries.add(PieEntry(100 * a.second / total, a.first))
        }
        // Data set
        val dataSet = PieDataSet(entries, "")
        dataSet.sliceSpace = 2f
        dataSet.valueTextSize = 12f
        dataSet.valueFormatter = PercentFormatter()
        //Colors
        val colors = ArrayList<Int>()
        colors.add(Color.GRAY)
        colors.add(Color.BLUE)
        colors.add(Color.RED)
        colors.add(Color.GREEN)
        colors.add(Color.CYAN)
        colors.add(Color.YELLOW)
        colors.add(Color.MAGENTA)
        dataSet.colors = colors
        val pieData = PieData(dataSet)
        data = pieData
        invalidate()
    }
}